src/vfio/vfio.c

```diff
@@ -0,0 +1,397 @@
+		errno = EMFILE;
+		return -1;
+	}
+
+	group->path = strdup(path);
+	if (!group->path)
+		// errno is set by strdup
+		return -1;
+
+	group->fd = vfio_group_open(group->path);
+	if (group->fd < 0) {
+		errno_saved = errno;
+		log_debug("failed to open vfio group\n");
+		goto free_group_path;
+	}
+
+	if (ioctl(group->fd, VFIO_GROUP_SET_CONTAINER, &vfio->fd)) {
+		errno_saved = errno;
+		log_debug("failed to add group to vfio container\n");
+		goto free_group_path;
+	}
+
+	return group->fd;
+
+free_group_path:
+	free((void *)group->path);
+
+	errno = errno_saved;
+
+	return -1;
+}
+
+int vfio_get_device_fd(struct vfio_container *vfio, const char *bdf)
+{
+	__autofree char *group = NULL;
+	int gfd, ret_fd;
+
+	group = pci_get_iommu_group(bdf);
+	if (!group) {
+		log_error("could not determine iommu group for device %s\n", bdf);
+		errno = EINVAL;
+		return -1;
+	}
+	log_info("vfio iommu group is %s\n", group);
+
+	gfd = vfio_get_group_fd(vfio, group);
+	if (gfd < 0)
+		return -1;
+
+	if (vfio_configure_iommu(vfio)) {
+		// JAG: Should we unset group VFIO_GROUP_UNSET_CONTAINER on failure?
+		log_error("failed to configure iommu: %s\n", strerror(errno));
+		return -1;
+	}
+
+	ret_fd = ioctl(gfd, VFIO_GROUP_GET_DEVICE_FD, bdf);
+	if (ret_fd < 0) {
+		log_debug("failed to get device fd\n");
+		return -1;
+	}
+
+	return ret_fd;
+}
+
+int vfio_init_container(struct vfio_container *vfio)
+{
+	vfio->fd = open("/dev/vfio/vfio", O_RDWR);
+	if (vfio->fd < 0) {
+		log_debug("failed to open vfio device\n");
+		return -1;
+	}
+
+	if (ioctl(vfio->fd, VFIO_GET_API_VERSION) != VFIO_API_VERSION) {
+		log_debug("invalid vfio version\n");
+		return -1;
+	}
+
+	if (!ioctl(vfio->fd, VFIO_CHECK_EXTENSION, VFIO_TYPE1_IOMMU)) {
+		log_debug("vfio type 1 iommu not supported\n");
+		return -1;
+	}
+	return 0;
+}
+
+int vfio_do_map_dma(struct vfio_container *vfio, void *vaddr, size_t len, uint64_t iova)
+{
+	struct vfio_iommu_type1_dma_map dma_map = {
+		.argsz = sizeof(dma_map),
+		.vaddr = (uintptr_t)vaddr,
+		.size  = len,
+		.iova  = iova,
+		.flags = VFIO_DMA_MAP_FLAG_READ | VFIO_DMA_MAP_FLAG_WRITE,
+	};
+
+	trace_guard(VFIO_IOMMU_MAP_DMA) {
+		trace_emit("vaddr %p iova 0x%" PRIx64 " len %zu\n", vaddr, iova, len);
+	}
+
+	if (!ALIGNED(((uintptr_t)vaddr | len | iova), __VFN_PAGESIZE)) {
+		log_debug("vaddr, len or iova not page aligned\n");
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (ioctl(vfio->fd, VFIO_IOMMU_MAP_DMA, &dma_map)) {
+		log_error("Error while mapping %s\n", strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+int vfio_do_unmap_dma(struct vfio_container *vfio, size_t len, uint64_t iova)
+{
+	struct vfio_iommu_type1_dma_unmap dma_unmap = {
+		.argsz = sizeof(dma_unmap),
+		.size = len,
+		.iova = iova,
+	};
+
+	trace_guard(VFIO_IOMMU_UNMAP_DMA) {
+		trace_emit("iova 0x%" PRIx64 " len %zu\n", iova, len);
+	}
+
+	if (ioctl(vfio->fd, VFIO_IOMMU_UNMAP_DMA, &dma_unmap)) {
+		log_error("Error while un-mapping %s\n", strerror(errno));
```

Comment:
In general, the style when using the logging functions is to just describe the event. I.e., here I'd just do

    log_error("failed to unmap: %s\n", strerror(errno));

The logging function will already output an`E` prefix to signify that it is an error.

Reply:
done